source/adapters/level_zero/v2/command_buffer.cpp

```diff
@@ -0,0 +1,290 @@
+  }
+
+  return {waitList.data(), static_cast<uint32_t>(numWaitEvents)};
+}
+
+ur_exp_command_buffer_handle_t_::ur_exp_command_buffer_handle_t_(
+    ur_context_handle_t context, ur_device_handle_t device,
+    v2::raii::command_list_unique_handle &&commandList,
+    const ur_exp_command_buffer_desc_t *desc)
+    : context(context), device(device),
+      zeCommandList(
+          std::forward<v2::raii::command_list_unique_handle>(commandList)),
+      isUpdatable(desc ? desc->isUpdatable : false) {
+  UR_CALL_THROWS(ur::level_zero::urContextRetain(context));
+  UR_CALL_THROWS(ur::level_zero::urDeviceRetain(device));
+}
+
+ur_exp_command_buffer_handle_t_::~ur_exp_command_buffer_handle_t_() {
+  // Release the memory allocated to the Context stored in the command_buffer
+  ur::level_zero::urContextRelease(context);
+
+  // Release the device
+  ur::level_zero::urDeviceRelease(device);
+
+  for (auto &associatedKernel : kernelsList) {
+    ur::level_zero::urKernelRelease(associatedKernel);
+  }
+}
+
+namespace ur::level_zero {
+
+/**
+ * Creates a L0 command list
+ * @param[in] context The Context associated with the command-list
+ * @param[in] device  The Device associated with the command-list
+ * @param[in] isUpdatable Whether the command-list should be mutable.
+ * @param[out] commandList The L0 command-list created by this function.
+ * @return UR_RESULT_SUCCESS or an error code on failure
+ */
+ur_result_t createMainCommandList(ur_context_handle_t context,
+                                  ur_device_handle_t device, bool isUpdatable,
+                                  ze_command_list_handle_t &commandList) {
+
+  using queue_group_type = ur_device_handle_t_::queue_group_info_t::type;
+  // that should be call to queue getZeOrdinal,
+  // but queue is not available while constructing buffer
+  uint32_t queueGroupOrdinal =
+      device->QueueGroup[queue_group_type::Compute].ZeOrdinal;
+
+  ZeStruct<ze_command_list_desc_t> zeCommandListDesc;
+  zeCommandListDesc.commandQueueGroupOrdinal = queueGroupOrdinal;
+
+  zeCommandListDesc.flags = ZE_COMMAND_LIST_FLAG_IN_ORDER;
+
+  ZeStruct<ze_mutable_command_list_exp_desc_t> zeMutableCommandListDesc;
+  if (isUpdatable) {
+    zeMutableCommandListDesc.flags = 0;
+    zeCommandListDesc.pNext = &zeMutableCommandListDesc;
+  }
+
+  ZE2UR_CALL(zeCommandListCreate, (context->getZeHandle(), device->ZeDevice,
+                                   &zeCommandListDesc, &commandList));
+
+  return UR_RESULT_SUCCESS;
+}
+
+ur_result_t
+urCommandBufferCreateExp(ur_context_handle_t context, ur_device_handle_t device,
+                         const ur_exp_command_buffer_desc_t *commandBufferDesc,
+                         ur_exp_command_buffer_handle_t *commandBuffer) {
+  try {
+    bool isUpdatable = commandBufferDesc && commandBufferDesc->isUpdatable;
+    checkImmediateAppendSupport(context);
+
+    if (isUpdatable) {
+      UR_ASSERT(context->getPlatform()->ZeMutableCmdListExt.Supported,
```

Comment:
Please don't use these "asserts". Do a normal if and return `UR_RESULT_ERROR_UNSUPPORTED_FEATURE`.
I just dislike overloading the term "assert" to mean fail with error.